drivers/metal/rendering_shader_container_metal.mm

```diff
@@ -0,0 +1,706 @@
+		res.features.mslVersionMinor = 2;
+	}
+
+	return &profiles.insert(key, res)->value;
+}
+
+Error RenderingShaderContainerMetal::compile_metal_source(const char *p_source, const StageData &p_stage_data, Vector<uint8_t> r_binary_data) {
+	String name(shader_name.ptr());
+	if (name.contains_char(':')) {
+		name = name.replace(":", "_");
+	}
+	Error r_error;
+	Ref<FileAccess> source_file = FileAccess::create_temp(FileAccess::ModeFlags::READ_WRITE,
+			name + "_" + itos(p_stage_data.hash.short_sha()),
+			"metal", false, &r_error);
+	if (r_error != OK) {
+		ERR_FAIL_V_MSG(r_error, "Unable to create temporary source file");
+	}
+	if (!source_file->store_buffer((const uint8_t *)p_source, strlen(p_source))) {
+		ERR_FAIL_V_MSG(ERR_CANT_CREATE, "Unable to write temporary source file");
+	}
+	source_file->flush();
+	Ref<FileAccess> result_file = FileAccess::create_temp(FileAccess::ModeFlags::READ_WRITE,
+			name + "_" + itos(p_stage_data.hash.short_sha()),
+			"metallib", false, &r_error);
+
+	if (r_error != OK) {
+		ERR_FAIL_V_MSG(r_error, "Unable to create temporary target file");
+	}
+
+	String sdk;
+	switch (device_profile->platform) {
+		case MetalDeviceProfile::Platform::macOS:
+			sdk = "macosx";
+			break;
+		case MetalDeviceProfile::Platform::iOS:
+			sdk = "iphoneos";
+			break;
+	}
+
+	// Build the metallib binary.
+	{
+		List<String> args{ "-sdk", sdk, "metal", "-O3" };
+		if (p_stage_data.is_position_invariant) {
+			args.push_back("-fpreserve-invariance");
+		}
+		args.push_back("-fmetal-math-mode=fast");
+		args.push_back(source_file->get_path_absolute());
+		args.push_back("-o");
+		args.push_back(result_file->get_path_absolute());
+		String r_pipe;
+		int exit_code;
+		Error err = OS::get_singleton()->execute("/usr/bin/xcrun", args, &r_pipe, &exit_code, true);
+		if (!r_pipe.is_empty()) {
+			print_line(r_pipe);
+		}
+		if (err != OK) {
+			ERR_PRINT(vformat("Metal compiler returned error code: %d", err));
+		}
+
+		if (exit_code != 0) {
+			ERR_PRINT(vformat("Metal compiler exited with error code: %d", exit_code));
+		}
+		int len = result_file->get_length();
+		if (len == 0) {
+			ERR_FAIL_V_MSG(ERR_CANT_CREATE, "Metal compiler created empty library");
+		}
+	}
+
+	// Strip the source from the binary.
+	{
+		List<String> args{ "-sdk", sdk, "metal-dsymutil", "--remove-source", result_file->get_path_absolute() };
+		String r_pipe;
+		int exit_code;
+		Error err = OS::get_singleton()->execute("/usr/bin/xcrun", args, &r_pipe, &exit_code, true);
+		if (!r_pipe.is_empty()) {
+			print_line(r_pipe);
+		}
+		if (err != OK) {
+			ERR_PRINT(vformat("metal-dsymutil tool returned error code: %d", err));
+		}
+
+		if (exit_code != 0) {
+			ERR_PRINT(vformat("metal-dsymutil Compiler exited with error code: %d", exit_code));
+		}
+		int len = result_file->get_length();
+		if (len == 0) {
+			ERR_FAIL_V_MSG(ERR_CANT_CREATE, "metal-dsymutil tool created empty library");
+		}
+	}
+
+	r_binary_data = result_file->get_buffer(result_file->get_length());
```

Comment:
This only worked for me (when backported to 4.4.x) if r_binary_data was passed as reference. It would also have less overhead than passing a copy of the Vector.

Reply:
Thanks @kisg – I've submitted a suggestion for Dario to resolve it.

Reply:
Should be resolved now.